clang/lib/Sema/Sema.cpp

```diff
@@ -1600,10 +1603,20 @@ class DeferredDiagnosticsEmitter
   void visitUsedDecl(SourceLocation Loc, Decl *D) {
     if (isa<VarDecl>(D))
       return;
-    if (auto *FD = dyn_cast<FunctionDecl>(D))
+    if (auto *FD = dyn_cast<FunctionDecl>(D)) {
+      Sema::DeviceDiagnosticReason SaveReason = RootReason;
```

Comment:
I'd prefer a RAII type for this functionality (and I think that is what I suggested earlier).  What do @AaronBallman / @elizabethandrews / @premanandrao think?